java/ingest/src/main/java/sleeper/ingest/impl/IngestCoordinatorFactory.java

```diff
@@ -0,0 +1,202 @@
+        String fileWriterType = instanceProperties.get(INGEST_PARTITION_FILE_WRITER_TYPE).toLowerCase(Locale.ROOT);
+        StandardIngestCoordinator.BackedBuilder ingestCoordinatorBuilder;
+        try {
+            if (recordBatchType.equals("arraylist")) {
+                ingestCoordinatorBuilder = StandardIngestCoordinator.builder().fromProperties(ingestProperties)
+                        .backedByArrayList()
+                        .maxNoOfRecordsInMemory((int) ingestProperties.getMaxInMemoryBatchSize())
+                        .maxNoOfRecordsInLocalStore(ingestProperties.getMaxRecordsToWriteLocally());
+            } else if (recordBatchType.equals("arrow")) {
+                ingestCoordinatorBuilder = StandardIngestCoordinator.builder().fromProperties(ingestProperties)
+                        .backedByArrow()
+                        .arrowBufferAllocator(bufferAllocator)
+                        .maxNoOfRecordsToWriteToArrowFileAtOnce(instanceProperties.getInt(ARROW_INGEST_MAX_SINGLE_WRITE_TO_FILE_RECORDS))
+                        .workingArrowBufferAllocatorBytes(instanceProperties.getLong(ARROW_INGEST_WORKING_BUFFER_BYTES))
+                        .minBatchArrowBufferAllocatorBytes(instanceProperties.getLong(ARROW_INGEST_BATCH_BUFFER_BYTES))
+                        .maxBatchArrowBufferAllocatorBytes(instanceProperties.getLong(ARROW_INGEST_BATCH_BUFFER_BYTES))
+                        .maxNoOfBytesToWriteLocally(ingestProperties.getMaxRecordsToWriteLocally());
+            } else {
+                throw new UnsupportedOperationException(String.format("Record batch type %s not supported", recordBatchType));
+            }
+            if (fileWriterType.equals("direct")) {
+                return ingestCoordinatorBuilder.buildDirectWrite(ingestProperties.getFilePrefix() + ingestProperties.getBucketName());
+            } else if (fileWriterType.equals("async")) {
+                if (!instanceProperties.get(FILE_SYSTEM).toLowerCase(Locale.ROOT).equals("s3a://")) {
+                    throw new UnsupportedOperationException("Attempting an asynchronous write to a file system that is not s3a://");
+                } else {
+                    return ingestCoordinatorBuilder.buildAsyncS3Write(ingestProperties.getBucketName(), internalS3AsyncClient);
+                }
+            } else {
+                throw new UnsupportedOperationException(String.format("Record batch type %s not supported", recordBatchType));
+            }
+        } finally {
+            if (s3AsyncClient == null && internalS3AsyncClient != null) {
+                internalS3AsyncClient.close();
+            }
+        }
+    }
+
+    public IngestRecordsFromIterator createIngestRecordsFromIterator(String tableName, Iterator<Record> recordIterator) {
+        return new IngestRecordsFromIterator(createIngestCoordinator(tableName), recordIterator);
```

Comment:
The IngestCoordinator was in a try-with-resources block before. We can't really guarantee it's closed this way. Maybe we should do the full ingest inside the factory and just return the IngestResult?